src/app/loadout-builder/process-worker/auto-stat-mod-utils.ts

```diff
@@ -0,0 +1,267 @@
+ * Auto mod assignment *must be* correct: If there is a way to pick stat mods and assign
+ * other mods and stat mods so that they fit on the items and the minimum stats are hit,
+ * LO must be able to find it. This is surprisingly tricky: Imagine we need one tier of
+ * strength and one tier of recovery. We could pick the mods in different ways:
+ *
+ * * [large recovery (4), large strength (3)]
+ * * [large recovery (4), small strength (1), small strength (1)]
+ * * [large strength (3), small recovery (2), small recovery (2)]
+ * * [small recovery (2), small recovery (2), small strength (1), small strength (1)]
+ *
+ * Each of these picks is pareto-optimal wrt. leftover energy. There exist armor + mod
+ * picks for which exactly one of these picks fits and no other pick can fit.
+ * This becomes even worse when LO has to assign bucket-independent mods like combat or
+ * activity mods, because there might be a permutation of BI mods that doesn't
+ * leave enough space for any of these picks and another permutation that does.
+ *
+ * What saves us here is that stat mods don't have an element or mod slot requirement, so we
+ * don't have to test all permutations of stat mods. Instead, we just sort the costs of each pick
+ * descending and compare with the leftover energy capacities, also sorted descending.
+ *
+ * When LO determines that a set is missing some stats, it asks this cache for picks of mods that
+ * generate these stats. `getViableGeneralModPicks` builds a list of possible picks.
+ * LO can then test these picks for every BI assignment it comes up with.
+ *
+ * Finally, what's perhaps interesting about this is that this already factors in stat mods the
+ * user picked themselves, but without the stats (because those are factored in externally in the base stats).
+ * E.g. if the user forces an intellect mod (cost 5), `getViableGeneralModPicks` will only generate picks with up
+ * to 4 extra mods, and the costs of every pick will include a 5 at the front. LO thus doesn't need to
+ * iterate over stat mod permutations and gets the check for free.
+ */
+export function createGeneralModsCache(
+  generalMods: ProcessMod[],
+  statOrder: ArmorStatHashes[],
+  autoStatMods: boolean
+): GeneralModsCache {
+  return {
+    generalModCosts: generalMods.map((mod) => mod.energy?.val || 0).sort(compareBy((x) => -x)),
+    statOrder,
+    autoStatMods,
+    cache: {},
+    cacheHits: 0,
+    cacheMisses: 0,
+    cacheSuccesses: 0,
+  };
+}
+
+export function getViableGeneralModPicks(
+  cache: GeneralModsCache,
+  neededStats: number[]
+): ModsPick[] {
+  if (!cache.autoStatMods) {
+    return [{ costs: cache.generalModCosts, modHashes: [] }];
+  }
+  // Divide by 5 and round up to the nearest integer, such that
+  // if we need [0]->0, [1-5]->1, [6-10]->2, [11-15]->3, ...
+  // This cache lookup is really hot code and can easily take up over
+  // 10% of the total processing time of the whole LO process.
```

Comment:
I suppose you could speed it up by unrolling loops and avoiding some conveniences like `map`. That really helped for the main LO loop.

Reply:
One thing to particularly note here is that the algorithm is sensitive to new array creation. If you can reuse an array rather than creating a new one each time when you are in the very bottom of the loops I found it helped a lot.